nixos/modules/services/networking/pleroma.xml

```diff
@@ -4,129 +4,185 @@
+
+# Configure web push notifications
+config :web_push_encryption, :vapid_details,
+  public_key: "&lt;the secret generated by pleroma_ctl&gt;",
+  private_key: "&lt;the secret generated by pleroma_ctl&gt;"
+
+# ... TO CONTINUE ...
+</programlisting>
+  Note that the lines of the same configuration group are comma separated (i.e. all the lines end with a comma, except the last one), so when the lines with passwords are added or removed, commas must be adjusted accordingly.</para>
+
+  <para>The service can be enabled with the usual
+<programlisting>
+<prompt>$ </prompt>nixos-rebuild switch
+</programlisting>
+  </para>
+  <para>The service is accessible only from the local <literal>127.0.0.1:4000</literal> port. It can be tested using a port forwarding like this
+<programlisting>
+<prompt>$ </prompt>ssh -L 4000:localhost:4000 myuser@example.net
+</programlisting>
+and then accessing <link xlink:href="http://localhost:4000">http://localhost:4000</link> from a web browser.</para>
+ </section>
+ <section xml:id="module-services-pleroma-admin-user">
+  <title>Creating the admin user</title>
+  <para>After Pleroma service is running, all <link xlink:href="https://docs-develop.pleroma.social/">Pleroma administration utilities</link> can be used. In particular an admin user can be created with
+<programlisting>
+<prompt>$ </prompt>pleroma_ctl user new &lt;nickname&gt; &lt;email&gt;  --admin --moderator --password &lt;password&gt;
+</programlisting>
+  </para>
+ </section>
+ <section xml:id="module-services-pleroma-nginx">
+  <title>Configuring Nginx</title>
+  <para>In this configuration, Pleroma is listening only on the local port 4000. Nginx can be configured as a Reverse Proxy, for forwarding requests from public ports to the Pleroma service. This is an example of configuration, using
+<link xlink:href="https://letsencrypt.org/">Let's Encrypt</link> for the TLS certificates
 <programlisting>
-    sudo -u postgres psql -f setup.psql
-</programlisting></para>
-   <para>In regard of the pleroma service configuration you also just generated, you'll need to split it in two parts. The "public" part, which do not contain any secrets and thus can be safely stored in the Nix store and its "private" counterpart containing some secrets (database password, endpoint secret key, salts, etc.).</para>
+security.acme = {
+  email = "root@example.net";
+  acceptTerms = true;
+};
 
-   <para>The public part will live in your NixOS machine configuration in the <link linkend="opt-services.pleroma.configs">services.pleroma.configs</link> option. However, it's up to you to upload the secret pleroma configuration to the path pointed by <link linkend="opt-services.pleroma.secretConfigFile">services.pleroma.secretConfigFile</link>. You can do that manually or rely on a third party tool such as <link xlink:href="https://github.com/DBCDK/morph">Morph</link> or <link xlink:href="https://github.com/NixOS/nixops">NixOps</link>.</para>
+services.nginx = {
```

Comment:
My only comment on this PR is that this can be added to the module at some point. Some services have an option to enable nginx that gives you common defaults.

it's definitely better to have this in the docs than nowhere at all.
Thank you for making the PR to add this.

Reply:
> Some services have an option to enable nginx that gives you common defaults.

We're not doing that on purpose. Half a decade of NixOS taught me the hard way it's usually not a good idea to perform some web-server autoconf magic in NixPkgs.

1. It's really hard to come up with a desirable default web conf. The webserver conf **heavily** depends on the deployment itself. As a end user, before using a module trying to configure your webserver, you'll have to answer some common questions such as "is the module doing some letsencrypt magic?" "Is this magic compatible with my current acme provisionning?" "I am using a different webserver, is this module going to break my overall config by making something listening to 443?". To answer that, you **have** to read the module definition. Which makes the module abstraction moot IMHO. It's fine to do that in your downstream Nix distribution, but it's not fine to do that in NixOS, we want to keep the distribution reasonably generic.
2. In this particular use case, we rely on some `extraConfig` options, specified as pure strings. Meaning you'll de-facto prevent the end-user to override these operations. Until the Nginx module adopts a RFC42-like style, we won't be able to provide a default Nginx conf for pleroma.

Reply:
I'm happy with that decision.
I personally use caddy, and for caddy I feel it makes sense to have some defaults. I might find out the hard way that it's not a good idea :-)
I like your idea about RFC42 style for a reverse proxy config, I'll think about what it means for caddy.